core-primitives/stf-state-handler/src/in_memory_state_file_io.rs

```diff
@@ -0,0 +1,367 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+#[cfg(feature = "sgx")]
+use std::sync::SgxRwLock as RwLock;
+
+#[cfg(feature = "std")]
+use std::sync::RwLock;
+
+use crate::{
+	error::{Error, Result},
+	file_io::StateFileIo,
+};
+use codec::Encode;
+use itp_types::ShardIdentifier;
+use std::{
+	collections::HashMap,
+	hash::Hasher as HasherTrait,
+	string::{String, ToString},
+	vec::Vec,
+};
+
+type StateHash = u64;
+type ShardDirectory<State> = HashMap<String, (StateHash, State)>;
+type ShardsRootDirectory<State> = HashMap<ShardIdentifier, ShardDirectory<State>>;
+
+/// State file I/O using (unencrypted) in-memory representation of the state files.
+/// Uses u64 hash type. Can be used as mock for testing.
+#[derive(Default)]
+pub struct InMemoryStateFileIo<State, Hasher>
+where
+	State: Clone + Default + Encode,
+	Hasher: HasherTrait + Clone + Default,
+{
+	emulated_file_directory: RwLock<ShardsRootDirectory<State>>,
+	hasher: Hasher,
+}
```

Comment:
In-memory implementation of the `StateFileIo` trait. Used for unit testing.
I looked at [Rust VFS](https://docs.rs/vfs/latest/vfs/index.html) for such an abstraction too (with `PhysicalFS` and `MemoryFs`), but it didn't seem to play nicely with our SGX/STD environment.